org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/notification/JavaBuilderState.java

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (c) 2013, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.common.types.ui.notification;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.internal.compiler.util.SimpleLookupTable;
+import org.eclipse.jdt.internal.core.JavaModelManager;
+import org.eclipse.jdt.internal.core.builder.State;
+import org.eclipse.jdt.internal.core.builder.StringSet;
+import org.eclipse.xtext.naming.QualifiedName;
+
+/**
+ * @since 2.5
+ * @author Anton Kosyakov - Initial contribution and API
+ */
+@SuppressWarnings("restriction")
+public class JavaBuilderState {
+
+	private static final Logger LOG = Logger.getLogger(JavaBuilderState.class);
+
+	private final State state;
+
+	private final IProject project;
+
+	private Integer buildNumber;
+
+	private SimpleLookupTable references;
+
+	private Long lastStructuralBuildTime;
+
+	private Set<QualifiedName> structurallyChangedTypes;
+
+	private JavaBuilderState(IProject project, State state) {
+		this.project = project;
+		this.state = state;
+	}
+
+	public IProject getProject() {
+		return project;
+	}
+
+	public static JavaBuilderState getLastBuiltState(IJavaElement javaElement) {
+		IJavaProject javaProject = javaElement.getJavaProject();
+		return javaProject == null ? null : JavaBuilderState.getLastBuiltState(javaProject.getProject());
+	}
+
+	public static JavaBuilderState getLastBuiltState(IProject project) {
+		final State state;
+		Object lastBuiltState = JavaModelManager.getJavaModelManager().getLastBuiltState(project, null);
+		if (lastBuiltState instanceof State) {
+			state = (State) lastBuiltState;
+		} else {
+			state = null;
+		}
+		return new JavaBuilderState(project, state);
+	}
+
+	public Long getLastStructuralBuildTime() {
+		if (lastStructuralBuildTime != null) {
+			return lastStructuralBuildTime;
+		}
+		if (state == null) {
+			return lastStructuralBuildTime = -1l;
```

Comment:
AFAIK capital letters are supposed for long literals. The new Eclipse also has a new warning for this.

Reply:
You are right. I am leaving a link here for others who might be interested in knowing the reason behind this recommendation (PS Integer Literals section) - https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html